### PR TITLE
Improve requeue lost messages

### PR DESCRIPTION
**Pull request by DocRoms**
### Why ?
I want to improve the requeue of old messages.

### Changes
* [x] Improve the requeue of old messages

### CODE REVIEW

File: src/MessageHandler/LostMessagesConsumer.php [hunk 1]
```diff
@@ -44,33 +44,46 @@ public function setMessageTtl(int $messageTtl): void
...
-        array_map(
-            [$this, 'requeueOldMessageFromList'],
-            iterator_to_array($this->queue->getWorkingLists($this->redisClient))
-        );
+        array_map([$this, 'requeueOldMessageFromList'], iterator_to_array($this->queue->getWorkingLists($this->redisClient)), [$secondTime]);
     }
 
-    protected function requeueOldMessageFromList(string $list)
+    /**
+     * Requeue old messages from $list, and remove them if $messageTTL or
```
Comment (Oliboy50) [inline src/MessageHandler/LostMessagesConsumer.php]: or what? 💃

File: src/MessageHandler/LostMessagesConsumer.php [hunk 2]
```diff
@@ -44,33 +44,46 @@ public function setMessageTtl(int $messageTtl): void
...
-                    $this->addMessageInQueueList($message);
+            $serializedMessage = $this->redisClient->rpoplpush($list, $list);
 
-                    continue;
-                }
+            $message = empty($serializedMessage) ? null : MessageEnvelope::unserializeMessage($serializedMessage);
+
+            if (!empty($message)) {
+                if ((time() - $message->getUpdatedAt()
+                                      ->format('U')) > $this->messageTtl && $this->redisClient->lrem($list, 0, $serializedMessage)) {
```
Comment (Oliboy50) [inline src/MessageHandler/LostMessagesConsumer.php]: this indent seems a bit messy (at least on GitHub)

File: src/MessageHandler/LostMessagesConsumer.php [hunk 3]
```diff
@@ -44,33 +44,46 @@ public function setMessageTtl(int $messageTtl): void
...
+                break;
+            }
 
             // check if message is old enough
-            $message = MessageEnvelope::unserializeMessage($serializedMessage);
-            if ((time() - $message->getUpdatedAt()->format('U')) > $this->messageTtl
-                && $this->redisClient->lrem($list, 0, $serializedMessage)) {
-                if ($message->getRetry() < $this->maxRetry) {
-                    $this->addMessageInQueueList($message);
+            $serializedMessage = $this->redisClient->rpoplpush($list, $list);
```
Comment (Oliboy50) [inline src/MessageHandler/LostMessagesConsumer.php]: I don't understand why this is done twice (here and in the `while` statement)

File: src/MessageHandler/LostMessagesConsumer.php (hunk 2 above)
Comment (DocRoms) [reply src/MessageHandler/LostMessagesConsumer.php]: It's strange, why the `make cs-fix` (php-cs-fixer)  not reformat that ?

File: src/MessageHandler/LostMessagesConsumer.php (hunk 3 above)
Comment (DocRoms) [reply src/MessageHandler/LostMessagesConsumer.php]: :bomb: me too. Thx for pre-review :smiley:

File: src/MessageHandler/LostMessagesConsumer.php [hunk 4]
```diff
@@ -44,33 +44,44 @@ public function setMessageTtl(int $messageTtl): void
...
+     * $messageTTL or time is exceed.
+     *
+     * @param string   $list
+     * @param int|null $secondTime
+     */
+    protected function requeueOldMessageFromList(string $list, ?int $secondTime = null)
     {
-        for ($i = 0, $n = $this->redisClient->llen($list); $i < $n; $i++) {
-            $serializedMessage = $this->redisClient->rpoplpush($list, $list);
+        do {
```
Comment (fabdsp) [inline src/MessageHandler/LostMessagesConsumer.php]: You should do a while and not a do while

File: src/MessageHandler/LostMessagesConsumer.php [hunk 5]
```diff
@@ -44,33 +44,44 @@ public function setMessageTtl(int $messageTtl): void
...
+                break;
+            }
 
             // check if message is old enough
-            $message = MessageEnvelope::unserializeMessage($serializedMessage);
-            if ((time() - $message->getUpdatedAt()->format('U')) > $this->messageTtl
-                && $this->redisClient->lrem($list, 0, $serializedMessage)) {
-                if ($message->getRetry() < $this->maxRetry) {
-                    $this->addMessageInQueueList($message);
+            $message = empty($serializedMessage) ? null : MessageEnvelope::unserializeMessage($serializedMessage);
```
Comment (fabdsp) [inline src/MessageHandler/LostMessagesConsumer.php]: If the message is empty, you continue the loop even if the list is empty - I think you should break the loop in this case and not wait until the max time.

Comment (fabdsp) [reply src/MessageHandler/LostMessagesConsumer.php]: And we need to delete empty message - or they will stay definitively in the list

File: src/MessageHandler/LostMessagesConsumer.php [hunk 6]
```diff
@@ -44,33 +44,50 @@ public function setMessageTtl(int $messageTtl): void
...
-            $serializedMessage = $this->redisClient->rpoplpush($list, $list);
+        while ($serializedMessage = $this->redisClient->rpoplpush($list, $list))
+        {
+            // Break the loop, if message is empty.
+            if (empty($serializedMessage)){
+                break;
+            }
+
+            // If maximum times is defined and time is exceed.
+            if (($secondTime !== null && time() >= (time() + $secondTime))) {
```
Comment (fabdsp) [inline src/MessageHandler/LostMessagesConsumer.php]: You should add or make a test for this "if" :trollface: 
Please read it again ;-)

Comment (fabdsp) [reply src/MessageHandler/LostMessagesConsumer.php]: And maybe better at the end of the loop

File: src/MessageHandler/LostMessagesConsumer.php [hunk 7]
```diff
@@ -44,31 +44,46 @@ public function setMessageTtl(int $messageTtl): void
...
-                && $this->redisClient->lrem($list, 0, $serializedMessage)) {
-                if ($message->getRetry() < $this->maxRetry) {
-                    $this->addMessageInQueueList($message);
-
-                    continue;
+        // Define time before the loop.
+        $time = time();
+
+        while ($serializedMessage = $this->redisClient->rpoplpush($list, $list)) {
+            // Check if message is old enough
```
Comment (fabdsp) [inline src/MessageHandler/LostMessagesConsumer.php]: Comment not at the right place

File: src/MessageHandler/LostMessagesConsumer.php [hunk 8]
```diff
@@ -44,36 +44,57 @@ public function setMessageTtl(int $messageTtl): void
...
+
+        shuffle($workingLists = iterator_to_array($this->queue->getWorkingLists($this->redisClient)));
+
+        foreach($workingLists as $workingList)
+        {
+            if ($restTime <= 0 ){
+                break;
+            }
+
+            $this->requeueOldMessageFromList($workingList, $restTime);
```
Comment (fabdsp) [inline src/MessageHandler/LostMessagesConsumer.php]: Where do you decrease $restTime ?

File: src/MessageHandler/LostMessagesConsumer.php [hunk 9]
```diff
@@ -44,36 +44,57 @@ public function setMessageTtl(int $messageTtl): void
...
-                if ($message->getRetry() < $this->maxRetry) {
-                    $this->addMessageInQueueList($message);
+        while (($maxMessages-- > 0)  && ($serializedMessage = $this->redisClient->rpoplpush($list, $list))) {
+            $message = empty($serializedMessage) ? null : MessageEnvelope::unserializeMessage($serializedMessage);
 
-                    continue;
-                }
+            if (!empty($message)) {
+                // Check if message is old enough
+                if (($time - $message->getUpdatedAt()->format('U')) > $this->messageTtl && $this->redisClient->lrem($list, 0, $serializedMessage)) {
```
Comment (fabdsp) [inline src/MessageHandler/LostMessagesConsumer.php]: time ??? $time is the max treatment time, nothing to compare the updateAt if the message ??

File: src/MessageHandler/LostMessagesConsumer.php [hunk 10]
```diff
@@ -44,36 +44,57 @@ public function setMessageTtl(int $messageTtl): void
...
+     * @param int|null $secondTime
      */
-    public function requeueOldMessages()
+    public function requeueOldMessages(?int $secondTime = null): void
     {
-        array_map(
-            [$this, 'requeueOldMessageFromList'],
-            iterator_to_array($this->queue->getWorkingLists($this->redisClient))
-        );
+        $restTime = $secondTime - (time() - $secondTime);
```
Comment (fabdsp) [inline src/MessageHandler/LostMessagesConsumer.php]: restTime will always be negative I think....

File: src/MessageHandler/LostMessagesConsumer.php [hunk 11]
```diff
@@ -44,36 +44,57 @@ public function setMessageTtl(int $messageTtl): void
...
 
-    protected function requeueOldMessageFromList(string $list)
+    /**
+     * Requeue old messages from $list, and remove them if
+     * $messageTTL or time is exceed.
+     *
+     * @param string   $list
+     * @param int $time
+     */
+    protected function requeueOldMessageFromList(string $list, int $time): void
```
Comment (fabdsp) [inline src/MessageHandler/LostMessagesConsumer.php]: $time => $maxExecTime ?

**Pull request merged**